lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,72 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Parse a Hash representing a JSON API document into an ActiveRecord-ready hash.
+          #
+          # @param [Hash] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          def parse(document, options = {})
+            fields = parse_fields(options[:fields])
+
+            hash = {}
+
+            primary_data = document.fetch('data', {})
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            hash.merge!(parse_attributes(primary_data['attributes'], fields))
+            hash.merge!(parse_relationships(primary_data['relationships'], fields))
+
+            hash
+          end
+
+          # @api private
```

Comment:
```ruby
# Given an Array
# For each element in the array, built a hash
# Then a symbol element is is used as both the key and value
# And a Hash element is merged into return value
# And all other values in the array are ignored
# @example
#    parse_fields_options(['a']) # => nil
#    parse_fields_options([:a]) # => { a: :a }
#    parse_fields_options([{'a' => :a]) # => { 'a' => :a }
#    parse_fields_options(nil) # => nil
#    parse_fields_options({}) # => nil
#    parse_fields_options(:a) # => nil
#    parse_fields_options([[:a]]) # => {}
#    parse_fields_options([:a, {a: :b}]) # => { a: :b }
#    parse_fields_options([{a: :b}, :a]) # => { a: :a }
#   etc
#   etc
# @params fields [Array<Hash, Symbol>] object to transform into a hash or nil
# @return [Hash] for array inputs
# @return [Nil] otherwise
```